HPCC-31832 Migrate and improve regression testing in GitHub Actions

Charan-Sharan opened the pull request:
initial commit for adapting bundleTest.sh to GitHub Actions

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

AttilaVamos commented on .github/workflows/bundleTest-thor.yml:
```diff
@@ -0,0 +1,238 @@
+          do
+              BUNDLE_NAME=${BUNDLES_TO_TEST[i]}
+              BUNDLE_REPO="https://github.com/hpcc-systems/${BUNDLES_TO_TEST[i]}.git"
+              INSTALL_CMD="ecl bundle install --update --force $ML_INSTALL_EXTRA ${BUNDLE_REPO}"
```
If the ML_INSTALL_EXTRA is not defined, please remove it.

AttilaVamos commented on .github/workflows/bundleTest-thor.yml:
```diff
@@ -0,0 +1,238 @@
+                  cat ${logfilename} >> /home/runner/HPCCSystems-regression/log/Failed_test.summary
+                  echo "uploadArtifact=true" >> $GITHUB_OUTPUT
+              fi
+              # Rename result log file to name of the bundle
```
If there is no error and no artifact upload then  is this renaming necessary?

AttilaVamos commented on .github/workflows/bundleTest-thor.yml:
```diff
@@ -0,0 +1,238 @@
+          
+          }
+
+          BUNDLES_TO_TEST=( "ML_Core" "PBblas" "GLM" "DBSCAN" "GNN" "LearningTrees" "TextVectors" "KMeans" "SupportVectorMachines" "LinearRegression" "LogisticRegression" )        
```
The 'BUNDLES_TO_TEST' was already defined, but yes in a different step. Define it once as a global and use anywhere where it needs (maintenance issue).

AttilaVamos commented on .github/workflows/bundleTest-thor.yml:
```diff
@@ -0,0 +1,238 @@
+              bundleName=$(basename $bundleName )         # remove path from $bundleName
+              
+              if [[ "$bundle" =~ "LearningTrees" ]]
+              then
```
Do you understand what this part does?

AttilaVamos commented on .github/workflows/bundleTest-thor.yml:
```diff
@@ -0,0 +1,238 @@
+              
+              pushd $bundleRunPath
+          
+              /opt/HPCCSystems/testing/regress/ecl-test run -t thor --config /opt/HPCCSystems/testing/regress/ecl-test.json --timeout 3600 -fthorConnectTimeout=3600 --pq 1 --ef ClassicTestModified.ecl,SVCTest.ecl,ClassificationTestModified.ecl
```
It would be better if these test names : 'ClassicTestModified.ecl,SVCTest.ecl,ClassificationTestModified.ecl' are stored in a global variable at the beginning (maintenance issue).

AttilaVamos commented on .github/workflows/bundleTest-thor.yml:
```diff
@@ -0,0 +1,238 @@
+      - name: Generate ZAP files
+        if: ${{ ! inputs.Generate-Zap == '' }} 
+        run: |  
+          IFS=' ' read -a ML_GENERATE_ZAP_FOR <<< ${{ inputs.Generate-Zap }}
```
If any ZAP file is forced to generate, then they should be add to the artifact and upload. Otherwise they are lost at the end of runner execution and no reason to generate them.

AttilaVamos commented on .github/workflows/bundleTest-thor.yml:
```diff
@@ -0,0 +1,238 @@
+
+      - name: Check for Core files
+        run: |
+          NUM_OF_ML_CORES=( $(sudo find /var/lib/HPCCSystems/ -iname 'core*' -mtime -1 -type f -exec printf "%s\n" '{}' \; ) )
```
Same is true for Core handling as was in ZAP file generation. One more extra thing, core file can created without any test failure (hidden crash). So, if any core file found, the trace file should be a part of the artifact and need to upload.

AttilaVamos commented on .github/workflows/bundleTest-thor.yml:
```diff
@@ -0,0 +1,238 @@
+                  lastSubdir=${base##*/}
+                  comp=${lastSubdir##my}
+          
+                  #sudo gdb --batch --quiet -ex "set interactive-mode off" -ex "thread apply all bt" -ex "quit" "/opt/HPCCSystems/bin/${comp}" $core | sudo tee "$core.trace" 2>&1
```
Please remove commented out code (everywhere).

AttilaVamos commented on .github/workflows/bundleTest-thor.yml:
```diff
@@ -0,0 +1,238 @@
+      - name: Get test stat
+        if: ${{ inputs.Get-Stat }}
+        run: |
+          ./QueryStat2.py -p /home/runner/HPCCSystems-regression/log/  -d '' -a --timestamp --compileTimeDetails 1 --graphTimings --allGraphItems --addHeader
```
If this part is running, the generated .csv and .cfg files should add to artifact and upload.

AttilaVamos reviewed (changes requested):
Your implementation is good. 
There are some maintenance and artifact handling issue what you need to address.

Charan-Sharan committed:
HPCC-31832 Migrate and improve regression testing in GitHub Actions

- Download and install the platform
- Install dependencies
- Set the environment, enable core generation, start the platform
- Install and test bundles
- Generate and check for ZAP report files
- Check Core files
- Create and upload the artifact

Signed-off-by: Charan-Sharan <charansharan2004@gmail.com>

AttilaVamos commented on .github/workflows/bundleTest-thor.yml:
```diff
@@ -0,0 +1,276 @@
+          for ((i=0; i<$BUNDLES_COUNT; i++))
+          do
+              BUNDLE_NAME=${BUNDLES_TO_TEST[i]}
+              BUNDLE_REPO="https://github.com/hpcc-systems/${BUNDLES_TO_TEST[i]}.git"
```
Please output the bundle name and it's repo name as well (as you already do bellow with the bundle version)

AttilaVamos reviewed (changes requested):
One small change needed.

Michael-Gardner commented on .github/workflows/bundleTest-thor.yml:
```diff
@@ -0,0 +1,276 @@
+          done< <(find . -iname 'ecl' -type d | sort )
+
+      - name: Generate ZAP files
+        if: ${{ ! inputs.Generate-Zap == '' }} 
```
all lower case for consistency in the code

Michael-Gardner commented on .github/workflows/bundleTest-thor.yml:
```diff
@@ -0,0 +1,276 @@
+        description: 'Run Query stat'
+        required: false
+        default: false
+      generate-Zap:
```
recommend you use all lower case just for consistency when you have variable names like this

Michael-Gardner commented:
Looks great so far, watching for the output here https://github.com/hpcc-systems/HPCC-Platform/actions/runs/9417604318/job/25949171789?pr=18695

Michael-Gardner reviewed (approved):
Finish up any of the changes that Attila suggested and I'm good with this going in.

AttilaVamos reviewed (approved):
I think it is ready to go.

Charan-Sharan commented:
@ghalliday Both reviewers has approved my code. Could you please merge this Pull Request?

ghalliday commented:
I have merged and marked the associated jira as closed - but I wonder if this PR should have referenced a subtask.
